### PR TITLE
Improve handling of developement version numbers

### PR DESCRIPTION
**Pull request by Luthaf**
The core idea here is that the repo contains version number of the "latest release", not the "next release"; and we dynamically change these if building a development version, transforming `a.b.c` to `a.(b+1).0-devX` on the fly.

This was prompted by the need to handle both a `-rcX` suffix and a `-devX` suffix on the version number, while making sure all version sort in the right way.

 I also think this makes more sense than what we where doing; and is what setuptools-scm seems to be doing as well.
Another advantage is that we only have to do one commit for the release, bumping the version number & adding a tag to it (instead of many commits to change the version number in case of release candidate, tag, and then another commit to update the version to the next release).

The second commit make it so we can handle different "in-development" status for the different packages. The idea is that we will not tag the repo with `v0.2.34`, but instead with `metatensor-core-v0.2.34`. Potentially a single commit can be tagged with both `metatensor-core-v0.2.34` and `metatensor-operations-v0.56.1`.

----
:books: Documentation preview :books:: https://metatensor--382.org.readthedocs.build/en/382/

### CODE REVIEW

**Comment by PicoCentauri** [comment]
I get the idea but have to think about this. Sounds good at first glance. This only works for the Python code though. The rust version will always be at the current release without the extra tag right?

**Comment by agoscinski** [comment]
> This only works for the Python code though. The rust version will always be at the current release without the extra tag right?

Doesn't it make sense to also do this for the rust package, so imitating it with existing tools
https://github.com/rust-lang/cargo/issues/6583 
https://crates.io/crates/cargo-edit

**Comment by Luthaf** [comment]
Currently the version seen by cargo/rustc is always set to the previous release version. Since there is no way to dynamically set the version for now, I don't think we can do much better. Also, it should not matter much, since the only way to use this code from other Rust project is to point to the git repo directly.

What we can and I think should do, is to change the version inside the CMakeLists.txt (which changes the version seen by users of the C and C++ API); and change the self-reported version in `mts_version()`.

**Comment by agoscinski** [comment]
> Currently the version seen by cargo/rustc is always set to the previous release version. Since there is no way to dynamically set the version for now, I don't think we can do much better. Also, it should not matter much, since the only way to use this code from other Rust project is to point to the git repo directly.

Can't we make a script for building the rust package where we set the version which is then used also in the `CMakeLists.txt`? Then its consistent in all packages.

**Comment by Luthaf** [comment]
Not really, since the rust version must be a string in the Cargo.toml file. And we don't want every commit to add some noise on this line, replacing `0.2.0.dev3` with `0.2.0.dev4` and so on. Also, there will only be some inconsistency for development versions, released version will have the same version number.

**Comment by agoscinski** [comment]
> Not really, since the rust version must be a string in the Cargo.toml file. And we don't want every commit to add some noise on this line, replacing 0.2.0.dev3 with 0.2.0.dev4 and so on. Also, there will only be some inconsistency for development versions, released version will have the same version number.

Ah okay `cargo set-version` just changes the `Cargo.toml` file and overwriting the version in the build.rs seems not possible yet https://github.com/rust-lang/cargo/issues/7554 One could temporary change the Cargo.toml for build, but that seems not so robust

**Review by agoscinski — COMMENTED** [review]
looks okay as far as I understand the code, but it is not a deep review

**Comment by Luthaf** [comment]
Updated the code again, switching back from bash to Python for the script which determines the number of commits since last tag since bash is a nightmare to fine on Windows. This adds a dependency on Python for pure C++ builds, but only in development mode, so I'm fine with it.

**Review by PicoCentauri — APPROVED** [review]
I am fine with this. TBH I think this is a bit over overengineering but we will see how stable this is. We can change it if we feel it is more stressful then helpful.

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -16,6 +16,8 @@ jobs:
     runs-on: ubuntu-20.04
     steps:
       - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply .github/workflows/docs.yml:20]: This is actually making the checkout slower ^_^. `fetch-depth: 0` means fetch all commits (default it to only fetch the latest one). We need all commits to be able to correctly determine the version.

**Pull request merged**